xen/arch/arm/coproc/coproc.c

```diff
@@ -0,0 +1,381 @@
+/*
+ * xen/arch/arm/coproc/coproc.c
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <xen/init.h>
+#include <xen/err.h>
+#include <xen/guest_access.h>
+
+#include "coproc.h"
+
+/* dom0_coprocs: comma-separated list of coprocs for domain 0. */
+static char __initdata opt_dom0_coprocs[128] = "";
+string_param("dom0_coprocs", opt_dom0_coprocs);
+
+/* The "framework's" global coprocs list is used to keep track
+ * of all coproc instances that have been registered in the framework */
+static LIST_HEAD(coprocs);
+static DEFINE_SPINLOCK(coprocs_lock);
+static int num_coprocs;
+
+int vcoproc_context_switch(struct vcoproc_instance *curr, struct vcoproc_instance *next)
+{
+    struct coproc_device *coproc;
+    int ret = 0;
+
+    if ( unlikely(curr == next) )
+        return 0;
+
+    /* It is no matter who will provide coproc */
+    coproc = next ? next->coproc : curr->coproc;
```

Comment:
The comment above is not really clear. I would suggest you to remove it or to make it more obvious.